nucypher/blockchain/eth/sol/source/contracts/Issuer.sol

```diff
@@ -11,35 +10,35 @@ import "zeppelin/token/ERC20/SafeERC20.sol";
 
 /**
 * @notice Contract for calculation of issued tokens
-* @dev |v2.1.1|
+* @dev |v3.1.1|
 */
 abstract contract Issuer is Upgradeable {
     using SafeERC20 for NuCypherToken;
-    using SafeMath for uint256;
     using AdditionalMath for uint32;
 
     event Burnt(address indexed sender, uint256 value);
     /// Issuer is initialized with a reserved reward
     event Initialized(uint256 reservedReward);
 
+    uint128 constant MAX_UINT128 = uint128(0) - 1;
+
     NuCypherToken public immutable token;
-    uint256 public immutable totalSupply;
+    uint128 public immutable totalSupply;
 
     uint256 public immutable miningCoefficient;
     uint256 public immutable lockedPeriodsCoefficient;
     uint32 public immutable secondsPerPeriod;
     uint16 public immutable rewardedPeriods;
 
-    uint16 public currentMintingPeriod;
     /**
     * Current supply is used in the mining formula and is stored to prevent different calculation
     * for stakers which get reward in the same period. There are two values -
     * supply for previous period (used in formula) and supply for current period which accumulates value
-    * before end of period. There is no order between them because of storage savings.
-    * So each time should check values of both variables.
+    * before end of period.
     */
-    uint256 public currentSupply1;
-    uint256 public currentSupply2;
+    uint128 public previousPeriodSupply;
+    uint128 public currentPeriodSupply;
```

Comment:
Previous optimization provided updating only one supply value and saved 20k gas in that way. Now we can update both fields without losing this savings because they are stored in one slot

Reply:
Agreed, this is easier to follow! Although the previous approach was pretty clever :) – so now the contract is no longer alternating between two variables to calculate `amount`, right? (previously `currentSupply1` and `currentSupply2`)

Reply:
yes, now only `previousPeriodSupply` is using